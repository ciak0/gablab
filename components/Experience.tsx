import {
  differenceInMonths, format,
} from 'date-fns';
import { FunctionComponent } from 'react';
import pluralize from '../utils/pluralize';

export interface ExperienceProps {
  from: Date,
  to?: Date,
  title: string,
  location: string,
  company: {
    name: string,
    href: string,
    color: string,
  },
  fullTime?: boolean,
  className?: string,
}

const Experience: FunctionComponent<ExperienceProps> = ({
  className,
  title,
  from,
  to,
  company,
  location,
  children,
  fullTime,
}) => {
  const months = differenceInMonths(to || new Date(), from);
  const years = Math.floor(months / 12);
  const monthsRem = Math.floor(months % 12);

  return (
    <div className={className}>
      <h2 className="font-bold">{title}</h2>
      <p>
        <a
          href={company.href}
          target="_blank"
          rel="noreferrer"
          className={`${company.color} underline`}
        >
          {company.name}
        </a>
        {' '}
        &mdash;
        {' '}
        {fullTime ? 'Full time' : '4 day week'}
      </p>
      <p className="text-neutral-500 text-xs">
        <time>
          {format(from, 'MMM yyyy')}
          {' â†’ '}
          {to && format(to, 'MMM yyyy')}
          {!to && 'Present'}
          {' '}
          {months > 12 && `(${years}${pluralize(' yr', years)} ${monthsRem}${pluralize(' mo', monthsRem)})`}
          {months <= 12 && `(${months}${pluralize(' mo', months)})`}
        </time>
      </p>
      <p className="text-neutral-500 text-xs">
        {location}
      </p>
      <div className="mt-2 font-sans text-sm">
        {children}
      </div>
    </div>
  );
};

export default Experience;
