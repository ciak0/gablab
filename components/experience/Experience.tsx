import type { FunctionComponent, ReactChild } from 'react';
import {
  differenceInMonths, format,
} from 'date-fns';
import pluralize from '../../utils/pluralize';
import Expandable from '../expandable/Expandable';
import LocalImage, { LocalImageProps } from '../local-image/LocalImage';

export interface ExperienceProps {
  className?: string,
  from: Date,
  to?: Date,
  title: string,
  location: string,
  company: {
    name: string,
    href: string,
    image: LocalImageProps['src'],
    office?: string,
  },
  summary: ReactChild,
  fullTime?: boolean,
}

const Experience: FunctionComponent<ExperienceProps> = ({
  className,
  title,
  from,
  to,
  company,
  location,
  fullTime,
  summary,
  children,
}) => {
  const months = differenceInMonths(to || new Date(), from) + 1;
  const years = Math.floor(months / 12);
  const monthsRem = Math.floor(months % 12);

  return (
    <div className={`${className} flex items-start`}>
      <div>
        <LocalImage
          className="rounded"
          src={company.image}
          width={48}
          height={48}
        />
      </div>
      <div className="flex-1 pl-2">
        <h2 className="font-bold">{title}</h2>
        <p>
          <a
            href={company.href}
            target="_blank"
            rel="noreferrer"
            className="underline"
          >
            {company.name}
          </a>
          {company.office && ` (${company.office})`}
          {' '}
          &mdash;
          {' '}
          {fullTime ? 'Full time' : '4 day week'}
        </p>
        <p className="text-neutral-500 text-sm">
          <time>
            {format(from, 'MMM yyyy')}
            {' → '}
            {to && format(to, 'MMM yyyy')}
            {!to && 'Present'}
            {' '}
            {months > 12 && `· ${years}${pluralize(' yr', years)} ${monthsRem}${pluralize(' mo', monthsRem)}`}
            {months <= 12 && `· ${months}${pluralize(' mo', months)}`}
          </time>
        </p>
        <p className="text-neutral-500 text-xs">
          {location}
        </p>
        <Expandable
          className="mt-2 font-sans text-sm"
          summary={summary}
        >
          {children}
        </Expandable>
      </div>
    </div>
  );
};

Experience.defaultProps = {
  className: '',
};

export default Experience;
