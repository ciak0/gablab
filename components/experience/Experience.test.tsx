import TestRenderer from 'react-test-renderer';
import Experience from './Experience';

describe('components/Experience', () => {
  it('should render correctly', () => {
    const tree = TestRenderer.create(
      <Experience
        title="Explosive researcher"
        location="US"
        from={new Date(2020, 0, 1)}
        to={new Date(2021, 11, 31)}
        company={{ name: 'ACME Co', href: 'http://acme.co', image: 'acme.jpg' }}
        summary="Design and development of Explosive Tennis Balls"
      >
        Tickle your friends! Surprise your opponents
      </Experience>,
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render correctly with full-time and office', () => {
    const tree = TestRenderer.create(
      <Experience
        title="Explosive researcher"
        location="US"
        fullTime
        from={new Date(2020, 0, 1)}
        to={new Date(2021, 11, 31)}
        company={{
          name: 'ACME Co',
          office: 'Explosive R&D',
          href: 'http://acme.co',
          image: 'acme.jpg',
        }}
        summary="Design and development of Explosive Tennis Balls"
      >
        Tickle your friends! Surprise your opponents
      </Experience>,
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
