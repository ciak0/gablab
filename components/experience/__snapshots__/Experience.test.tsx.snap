// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Experience should render correctly 1`] = `
<div
  className=" flex items-start"
>
  <div>
    <span
      style={
        Object {
          "background": "none",
          "border": 0,
          "boxSizing": "border-box",
          "display": "inline-block",
          "height": "initial",
          "margin": 0,
          "maxWidth": "100%",
          "opacity": 1,
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "width": "initial",
        }
      }
    >
      <span
        style={
          Object {
            "background": "none",
            "border": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": "initial",
            "margin": 0,
            "maxWidth": "100%",
            "opacity": 1,
            "padding": 0,
            "width": "initial",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDgiIGhlaWdodD0iNDgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
          style={
            Object {
              "background": "none",
              "border": 0,
              "display": "block",
              "height": "initial",
              "margin": 0,
              "maxWidth": "100%",
              "opacity": 1,
              "padding": 0,
              "width": "initial",
            }
          }
        />
      </span>
      <img
        className="rounded"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
      <noscript>
        <img
          className="rounded"
          data-nimg="intrinsic"
          decoding="async"
          loading="lazy"
          src="acme.jpg"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </noscript>
    </span>
  </div>
  <div
    className="flex-1 pl-2"
  >
    <h2
      className="font-bold"
    >
      Explosive researcher
    </h2>
    <p>
      <a
        className="underline"
        href="http://acme.co"
        rel="noreferrer"
        target="_blank"
      >
        ACME Co
      </a>
       
      —
       
      4 day week
    </p>
    <p
      className="text-neutral-500 text-sm"
    >
      <time>
        Jan 2020
         → 
        Dec 2021
         
        · 2 yrs 0 mo
      </time>
    </p>
    <p
      className="text-neutral-500 text-xs"
    >
      US
    </p>
    <div
      className="mt-2 font-sans text-sm"
    >
      Design and development of Explosive Tennis Balls
      <button
        className="block mt-4 text-sm text-neutral-500"
        onClick={[Function]}
        type="button"
      >
        ...see more
      </button>
    </div>
  </div>
</div>
`;

exports[`components/Experience should render correctly with full-time and office 1`] = `
<div
  className=" flex items-start"
>
  <div>
    <span
      style={
        Object {
          "background": "none",
          "border": 0,
          "boxSizing": "border-box",
          "display": "inline-block",
          "height": "initial",
          "margin": 0,
          "maxWidth": "100%",
          "opacity": 1,
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "width": "initial",
        }
      }
    >
      <span
        style={
          Object {
            "background": "none",
            "border": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": "initial",
            "margin": 0,
            "maxWidth": "100%",
            "opacity": 1,
            "padding": 0,
            "width": "initial",
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDgiIGhlaWdodD0iNDgiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
          style={
            Object {
              "background": "none",
              "border": 0,
              "display": "block",
              "height": "initial",
              "margin": 0,
              "maxWidth": "100%",
              "opacity": 1,
              "padding": 0,
              "width": "initial",
            }
          }
        />
      </span>
      <img
        className="rounded"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style={
          Object {
            "border": "none",
            "bottom": 0,
            "boxSizing": "border-box",
            "display": "block",
            "height": 0,
            "left": 0,
            "margin": "auto",
            "maxHeight": "100%",
            "maxWidth": "100%",
            "minHeight": "100%",
            "minWidth": "100%",
            "objectFit": undefined,
            "objectPosition": undefined,
            "padding": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 0,
          }
        }
      />
      <noscript>
        <img
          className="rounded"
          data-nimg="intrinsic"
          decoding="async"
          loading="lazy"
          src="acme.jpg"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
      </noscript>
    </span>
  </div>
  <div
    className="flex-1 pl-2"
  >
    <h2
      className="font-bold"
    >
      Explosive researcher
    </h2>
    <p>
      <a
        className="underline"
        href="http://acme.co"
        rel="noreferrer"
        target="_blank"
      >
        ACME Co
      </a>
       (Explosive R&D)
       
      —
       
      Full time
    </p>
    <p
      className="text-neutral-500 text-sm"
    >
      <time>
        Jan 2020
         → 
        Dec 2021
         
        · 2 yrs 0 mo
      </time>
    </p>
    <p
      className="text-neutral-500 text-xs"
    >
      US
    </p>
    <div
      className="mt-2 font-sans text-sm"
    >
      Design and development of Explosive Tennis Balls
      <button
        className="block mt-4 text-sm text-neutral-500"
        onClick={[Function]}
        type="button"
      >
        ...see more
      </button>
    </div>
  </div>
</div>
`;
